### ISSUE TITLE
nvec should be a 1d array (or list) of ints

### ISSUE THREAD

**Issue by ne0hit** [bzier/gym-mupen64plus#40]
Hello bzier,  I ran into your project when I was digging about Tensorflow/Tensorkart so I decided to give it a try and see how it works. After some changes (bold text) suggested in previous comments. I was able to record, prepare samples and finally try to play the trained agent but now I am getting an error so maybe you might know whats happening. Here the changes that I have added to my environment some screenshots and a video.

If i dont use EmptyImage and EmptyBitmap wx does not work.
record.py
Images

    img = wx.**EmptyImage**(320,240)
    self.image_widget = wx.StaticBitmap(self.img_panel, wx.ID_ANY, wx.**EmptyBitmap**(320,240))
    bmp = wx.**EmptyBitmap**(Screenshot.SRC_W, Screenshot.SRC_H)

I had to disable XVFB from the config file otherwise I am not seeing anything on my desktop, for some reason it is not working for me.
config.yml (gym_mupen64plus)
XVFB config:

USE_XVFB: false
XVFB_CMD: Xvfb
TMP_DIR: /dev/shm
VGLRUN_CMD: vglrun

Finally when I run the play.py I am getting an assertion error.

![screenshot from 2018-03-19 15-13-38](https://user-images.githubusercontent.com/37481118/37619548-172045d6-2bba-11e8-8c1a-0f9eade3588d.png)

Here a video when the issue happens. Just before the race starts.

[AssertionError nvec.mp4.zip](https://github.com/bzier/gym-mupen64plus/files/1826760/AssertionError.nvec.mp4.zip)

Hope you could help and give some advise.

Best Regards!

Oscar

**Comment by bzier** [comment]
Hi Oscar,

Thanks for the issue (and welcome to GitHub). I'll try to take a closer look at it this evening and see what suggestions I may have for you.
>For traceability, linking to [this](https://github.com/kevinhughes27/TensorKart/issues/59) issue. Depending on where the problem lies, we may end up having the discussion on that thread instead. Otherwise, just getting them linked together for other people to be able to find the discussion.

-Brian

**Comment by ne0hit** [comment]
Thanks, meantime I will try to investigate by my own.

**Comment by bzier** [comment]
@ne0hit,
Ok, so I have a couple comments and an answer for the issue (I hope).

## Comments:
When using `XVFB`, the emulator actually renders to a virtual framebuffer, so nothing will show on the screen unless `env.render()` is explicitly called. Rendering is not available until the race has been set up, so the N64 startup and Mario Kart menu navigation will happen with nothing showing on screen. For TensorKart, once the race is ready to start, the screen will be rendered (and on each subsequent step). See the TensorKart code [here](https://github.com/kevinhughes27/TensorKart/blob/ed1ae0c787c89ca9fce46cd9ae8b6e06c14e766e/play.py#L63) and [here](https://github.com/kevinhughes27/TensorKart/blob/ed1ae0c787c89ca9fce46cd9ae8b6e06c14e766e/play.py#L75).

Based on the error that you are encountering, I believe it would happen regardless of whether or not you are using `XVFB`. However, because the error is occurring during the call to `gym.make()` when setting up the environment action space, this is prior to anything being rendered when using `XVFB`, so you wouldn't see anything at all. If you encounter different behavior or errors when you have `XVFB` turned on, please let me know, as that may be a separate problem.

When `XVFB` is turned off, you will need to ensure that the emulator window is in the proper position for the screenshots to be captured, otherwise the agent may 'see' other elements of your screen instead (e.g. desktop wallpaper, terminal/browser windows, etc). This is also configured in the `config.yml` file, with the values for `OFFSET_X` and `OFFSET_Y`. I noticed that in your video, you re-positioned the emulator window which would likely cause these offsets to be incorrect. That may have only been for the sake of the video, but I thought was worth pointing out.

## Answer:
Ok, now on to the problem... Turns out that there was a [breaking change](https://github.com/openai/gym/pull/836) made to the `gym` library a couple months ago which is affecting the initialization of the environment. The `gym` library was at version `0.7.4` for a long time (almost a year) and was working fine. The latest version (`0.9.6` - see [here](https://github.com/openai/gym/releases)) includes the breaking change. Ultimately, I will need to update my project to account for these breaking changes. However, in the meantime, you should be able to get up-and-running by changing two files in your `/home/user/workspace/mupen64/gym-mupen64plus/` directory:
- `environment.yml` - [line 16](https://github.com/bzier/gym-mupen64plus/blob/3551409db59ccba1912c004c1612b2b1aa2ecc8d/environment.yml#L16)
- `setup.py` - [line 5](https://github.com/bzier/gym-mupen64plus/blob/3551409db59ccba1912c004c1612b2b1aa2ecc8d/setup.py#L5)

Each of them should have the `gym` library set explicitly to use the older version (`gym=0.7.4`), which should then behave as expected. You may need to re-run the setup process after making this change:
`pip install -e .`

---

Please let me know if you have any further questions, or if the work-around doesn't do the trick for you. Thank you for reporting the issue, since (as far as I know) you're the first to bump into this. I was unaware that the `gym` library had been updated recently, and apparently have some work to do to bring the project up-to-speed with the latest & greatest.

**Comment by bzier** [comment]
Updated title to reflect the error message encountered, to aid in future searches

**Comment by ne0hit** [comment]
Thanks for your reply Brian, now I now why I was not seeing anything using XVBF, btw sorry for the delay but it might take a few hours before you can see a reply back from me since you are a few hours behind my timezone (Germany) I think. I will try again with the changes you mentioned and then I will write back to you with the outcome. Regarding the OFFSET_X/Y variables I set them with a value of zero in config file which means that I should put the mupen* emulator screen in the top left corner of my desktop screen am I right?. Being said that for the sake of the video, yes, I did re-positioned the emulator window.

**Comment by bzier** [comment]
No problem with the delay. That's part of the fun with these open source projects, getting to work with people all around the world. I'm in the U.S. Pacific timezone.

Regarding the offsets, you are correct, if you set them to 0,0 that should be the upper left corner of the screen. Just keep in mind that it may include menu/title bars as well. The simpler option is to use `XVFB` to guarantee that you are just capturing the emulator screen itself and nothing else.

Please do keep me updated with any progress or additional issues.

**Comment by ne0hit** [comment]
I am back, I added the changes but now I am stuck running the pip install, is worth to mention a couple of details. I am using two "=" operators within the setup.py file otherwise the script throws an error claiming that the command expects a string :/. the environment.yml file was left with just one operator so I am not sure if this should be updated to use == but I dont think so since there another dependencies there using the =. I ran the pip as sudo -H pip install -e . because I was seeing this error:

The directory '/home/user/.cache/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
The directory '/home/user/.cache/pip' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.

but thats the least of the problems the issue now is that the pip install is kind of stuck "Building wheels for collected packages: wxPython
  Running setup.py bdist_wheel for wxPython

To be honest I can't remember if this takes while. I think I ran the command late at night in like zombie mode. haha

**Comment by bzier** [comment]
Changing to `==` makes sense. I typed the info above from memory, so I may have messed up the syntax for the `setup.py` file. The `environment.yml` file is only applicable if you are using a [conda environment](https://github.com/bzier/gym-mupen64plus#method-2-installing-in-a-conda-environment) so if you're not, that file shouldn't make a difference.

Yes, that `wxPython` installation takes forever (hours in my case). However, I'm curious why it is attempting to build/install `wxPython` here. For this project, I removed that dependency a while back. It is still a dependency in the TensorKart project, but not in this one. Just make sure you're executing the `pip install -e .` command in the `gym-mupen64plus` directory. *(The addition of `sudo -H` should be fine, I think)*

**Comment by ne0hit** [comment]
I did a fresh install of gym-mupenplus64 and now its working!, as you said the issue was the gym version and I don't know why my setup.py had the inclusion of wxptyhon so thats the reason of the fresh install to avoid any mis-copies. I will try to understand how everything works for now. Thanks Brian. Btw if I have any question, not an issue exactly should I submit an issue or how do I proceed?

Regards!

**Comment by bzier** [comment]
Awesome! I'm glad to hear you got it working 👍
Any questions you have are fine to submit as new issues. I can just add a 'question' label to them for organization/searchability.

I'll go ahead and close this one since you were able to work through to a solution. I'll probably create a new issue to track upgrading this project for compatibility with the new `gym` changes.
